### PR TITLE
[DO NOT MERGE] Readable quickfix: ptxas shared memory error

### PR DESCRIPTION
**Pull request by emergenz**
A readable (and much more elegant) solution to overcoming the ptxas shared memory error (see https://github.com/p-doom/reward-redistribution/pull/9 for the unreadable version and additional context).

This PR is still not to be merged in its current state because we use `jax.pure_callback` to overcome the error. However, as a side-effect `pure_callback` moves data between xPU and CPU, hence slowing down the entire implementation (even though the error only occurs in the case `disable_autoreset=False` and `jax.debug.breakpoint()` set).

### CODE REVIEW

**Comment by emergenz** [comment]
Wrong PR target, sorry about that :)
